### PR TITLE
Traefik v2, TraefikProxy v1

### LINKED ISSUE #513: Specify ciphersuites in TLS config
We should probably explicitly specify our ciphersuites in traefik config.

----

Just as an FYI for anyone that may come across a similar issue, our VM team flagged us on use of an insecure cipher suite related to CVE-2016-2183 even after setting the minimum TLS version to v1.2, In our case we specified several suites in the traefik.toml config as a fix, e.g. 

```
minVersion = "VersionTLS12"
  cipherSuites = [
    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_RSA_WITH_AES_128_CBC_SHA",
    "TLS_RSA_WITH_AES_256_CBC_SHA"
  ]
```

_Originally posted by @aye-aye-aye in https://github.com/jupyterhub/the-littlest-jupyterhub/issues/488#issuecomment-587454858_

### PR DESCRIPTION
**Pull request by minrk**
Updates:

- jupyterhub-traefik-proxy to 1.0
- traefik to 2.10

traefik v2 totally changed their config file format, so several config options needed to move around, but ultimately the behavior is unchanged. jupyterhub-traefik-proxy 1.0 is required to use traefik v2 due to the config and API changes. Some configuration is no longer required as jupyterhub-traefik-proxy writes initial dynamic config (api route, auth) during its startup, instead of requiring to duplicate that info to the traefik config file.

This is a breaking change for the rare cases where users configured traefik or TraefikProxy directly, since both are major changes, but for most users the change should not be noticed (🤞) except as a performance boost when there are many users.

closes #513

### CODE REVIEW

**Comment by consideRatio** [comment]
Wieee tests are passing, ping me for review when its ready for review @minrk!

**Comment by minrk** [comment]
Tests are passing and I did a test deployment to verify that the letsencrypt config works. Good to go from me!

**Comment by consideRatio** [comment]
@minrk can you update title, PR description, and labels?

If there is breaking changes, can you clarify them in the PR description in a dedicated section as well?

My rough understanding is that we are updating the traefik-proxy, and setting it up to work against traefik v2 instead of v1.7.33. As part of this, we update misc config of traefik-proxy as well that ships with the TLJH distribution.

**Comment by minrk** [comment]
It's _possible_ the 0.2.0 upgrade test failed because of https://github.com/jupyterhub/the-littlest-jupyterhub/pull/895. I don't understand the error message well enough to be sure, but I ran a test deployment and can confirm that without the PATH change upgrading from 0.2.0 to jupyterhub 4 does not work, so I don't really understand why the test are passing most of the time. It could be that the initial 0.2.0 setup doesn't let old jupyterhub get fully set up enough.

File: tljh/traefik-dynamic.toml.tpl [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# traefik.toml dynamic config (mostly TLS)
+# dynamic config in the static config file will be ignored
+{% if https['enabled'] %}
+[tls]
+  [tls.options.default]
+  minVersion = "VersionTLS12"
+
```
Comment (MridulS) [inline tljh/traefik-dynamic.toml.tpl:7]:
```suggestion
  minVersion = "VersionTLS12"
  cipherSuites = [
    "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305",
    "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305"
    ]
```

while this is being fixed we should probably also sync this up with [z2jh config](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/73d4f182cc75d3761af1210ac53e81a17e400fe3/jupyterhub/templates/proxy/autohttps/_configmap-dynamic.yaml#LL101C1-L106C47). This also resolves https://github.com/jupyterhub/the-littlest-jupyterhub/issues/513

**Comment by minrk** [comment]
All green!

The intermittent failure in the integration test seems to be increasing in frequency. I still don't really understand the cause.

File: tljh/traefik-dynamic.toml.tpl [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# traefik.toml dynamic config (mostly TLS)
+# dynamic config in the static config file will be ignored
+{% if https['enabled'] %}
```
Comment (consideRatio) [inline tljh/traefik-dynamic.toml.tpl:3]:
```suggestion
{%- if https['enabled'] %}
```

File: tljh/traefik-dynamic.toml.tpl [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  minVersion = "VersionTLS12"
+  cipherSuites = [
+    "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
+    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
+    "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
+    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
+    "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305",
+    "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305",
+  ]
+  {% if https['tls']['cert'] -%}
```
Comment (consideRatio) [inline tljh/traefik-dynamic.toml.tpl:15]:
```suggestion
  {%- if https['tls']['cert'] %}
```

File: tljh/traefik-dynamic.toml.tpl [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  [tls.stores.default.defaultCertificate]
+    certFile = "{{ https['tls']['cert'] }}"
+    keyFile = "{{ https['tls']['key'] }}"
+  {%- endif %}
+
+  {% if https['letsencrypt']['email'] and https['letsencrypt']['domains'] -%}
```
Comment (consideRatio) [inline tljh/traefik-dynamic.toml.tpl:21]:
```suggestion
  {%- if https['letsencrypt']['email'] and https['letsencrypt']['domains'] %}
```

File: tljh/traefik-dynamic.toml.tpl [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  resolver = "letsencrypt"
+    [tls.stores.default.defaultGeneratedCert.domain]
+    main = "{{ https['letsencrypt']['domains'][0] }}"
+    sans = [
+      {% for domain in https['letsencrypt']['domains'][1:] -%}
+      "{{ domain }}",
+      {%- endfor %}
+    ]
+  {%- endif %}
+{% endif %}
```
Comment (consideRatio) [inline tljh/traefik-dynamic.toml.tpl:32]:
```suggestion
{%- endif %}
```

File: tljh/traefik.toml.tpl [hunk 6]
```diff
@@ -1,74 +1,61 @@
...
-  address = ":{{http['port']}}"
-  {% if https['enabled'] %}
-    [entryPoints.http.redirect]
-    entryPoint = "https"
-  {% endif %}
+  address = ":{{ http['port'] }}"
+  [entryPoints.http.transport.respondingTimeouts]
+  idleTimeout = "10m"
 
   {% if https['enabled'] %}
```
Comment (consideRatio) [inline tljh/traefik.toml.tpl:29]:
```suggestion
  {%- if https['enabled'] %}
```

File: tljh/traefik.toml.tpl [hunk 7]
```diff
@@ -1,74 +1,61 @@
...
-
 [entryPoints]
   [entryPoints.http]
-  address = ":{{http['port']}}"
-  {% if https['enabled'] %}
-    [entryPoints.http.redirect]
-    entryPoint = "https"
-  {% endif %}
+  address = ":{{ http['port'] }}"
+  [entryPoints.http.transport.respondingTimeouts]
```
Comment (consideRatio) [inline tljh/traefik.toml.tpl:26]:
```suggestion

  [entryPoints.http.transport.respondingTimeouts]
```

File: tljh/traefik.toml.tpl [hunk 8]
```diff
@@ -1,74 +1,61 @@
...
-  {% endif %}
-  {% endif %}
-  [entryPoints.auth_api]
-  address = "127.0.0.1:{{traefik_api['port']}}"
-  [entryPoints.auth_api.whiteList]
-  sourceRange = ['{{traefik_api['ip']}}']
-  [entryPoints.auth_api.auth.basic]
-  users = ['{{ traefik_api['basic_auth'] }}']
+  address = ":{{ https['port'] }}"
+  [entryPoints.https.http.tls]
```
Comment (consideRatio) [inline tljh/traefik.toml.tpl:36]:
```suggestion

  [entryPoints.https.http.tls]
```

File: tljh/traefik.toml.tpl [hunk 9]
```diff
@@ -1,74 +1,61 @@
...
+  options = "default"
 
-[wss]
-protocol = "http"
+  [entryPoints.https.transport.respondingTimeouts]
+  idleTimeout = "10m"
+  {% endif %}
```
Comment (consideRatio) [inline tljh/traefik.toml.tpl:41]:
```suggestion
  {%- endif %}
```

File: tljh/traefik.toml.tpl [hunk 10]
```diff
@@ -1,74 +1,61 @@
...
-[api]
-dashboard = true
-entrypoint = "auth_api"
+  [entryPoints.auth_api]
+  address = "localhost:{{ traefik_api['port'] }}"
 
-{% if https['enabled'] and https['letsencrypt']['email'] %}
-[acme]
-email = "{{https['letsencrypt']['email']}}"
+{% if https['enabled'] and https['letsencrypt']['email'] and https['letsencrypt']['domains'] %}
```
Comment (consideRatio) [inline tljh/traefik.toml.tpl:46]:
```suggestion
{%- if https['enabled'] and https['letsencrypt']['email'] and https['letsencrypt']['domains'] %}
```

File: tljh/traefik.toml.tpl [hunk 11]
```diff
@@ -1,74 +1,61 @@
...
 storage = "acme.json"
-entryPoint = "https"
-  [acme.httpChallenge]
-  entryPoint = "http"
-
-{% for domain in https['letsencrypt']['domains'] %}
-[[acme.domains]]
-  main = "{{domain}}"
-{% endfor %}
+{% if https['letsencrypt']['staging'] -%}
```
Comment (consideRatio) [inline tljh/traefik.toml.tpl:50]:
```suggestion
{%- if https['letsencrypt']['staging'] %}
```

File: tljh/traefik.toml.tpl [hunk 12]
```diff
@@ -1,74 +1,61 @@
...
+caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
+{%- endif %}
+[certificatesResolvers.letsencrypt.acme.tlsChallenge]
 {% endif %}
```
Comment (consideRatio) [inline tljh/traefik.toml.tpl:54]:
```suggestion
{%- endif %}
```

**Review by consideRatio — APPROVED** [review]
Wieee! This LGTM @minrk - if you agree on 324ded0003c0f0ae38188219ea8c5c95bec2bfe3 that I committed, then let's go for a merge!

**Comment by minrk** [comment]
Yes, thanks! I still don't have good intuition for chomping.

**Comment by consideRatio** [comment]
:heart: :tada: wieeeeeeeeeeeeeeeee!!!!

Thank you @minrk for your efforts into this!!!!

**Pull request merged**